server/src/main/java/org/elasticsearch/health/node/selection/HealthNodeSelectorTaskExecutor.java

```diff
@@ -0,0 +1,192 @@
+        this.taskStarter = this::startTask;
+        this.shutdownListener = this::shuttingDown;
+        if (ENABLED_SETTING.get(settings)) {
+            clusterService.addListener(taskStarter);
+            clusterService.addListener(shutdownListener);
+        }
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::enable);
+    }
+
+    private void enable(boolean enabled) {
+        if (enabled) {
+            clusterService.addListener(taskStarter);
+            clusterService.addListener(shutdownListener);
+        } else {
+            clusterService.removeListener(taskStarter);
+            clusterService.removeListener(shutdownListener);
+            abortTaskIfApplicable("disabling health node via '" + ENABLED_SETTING.getKey() + "'");
+        }
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, HealthNodeSelectorTaskParams params, PersistentTaskState state) {
+        HealthNodeSelector healthNodeSelector = (HealthNodeSelector) task;
+        currentTask.set(healthNodeSelector);
+        String nodeId = clusterService.localNode().getId();
+        logger.info("Node [" + nodeId + "] is selected as the current health node.");
+    }
+
+    @Override
+    protected HealthNodeSelector createTask(
+        long id,
+        String type,
+        String action,
+        TaskId parentTaskId,
+        PersistentTasksCustomMetadata.PersistentTask<HealthNodeSelectorTaskParams> taskInProgress,
+        Map<String, String> headers
+    ) {
+        return new HealthNodeSelector(id, type, action, getDescription(taskInProgress), parentTaskId, headers);
+    }
+
+    /**
+     * Returns the node id from the eligible health nodes
+     */
+    @Override
+    public PersistentTasksCustomMetadata.Assignment getAssignment(
+        HealthNodeSelectorTaskParams params,
+        Collection<DiscoveryNode> candidateNodes,
+        ClusterState clusterState
+    ) {
+        DiscoveryNode discoveryNode = selectLeastLoadedNode(clusterState, candidateNodes, DiscoveryNode::canContainData);
+        if (discoveryNode == null) {
+            return NO_NODE_FOUND;
+        } else {
+            return new PersistentTasksCustomMetadata.Assignment(discoveryNode.getId(), "");
+        }
+    }
+
+    // visible for testing
+    void startTask(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            // If the node is not a master node or if the task already exists this node should not start the task anymore
+            if (event.state().nodes().getLocalNode().isMasterNode() == false || HealthNodeSelector.findTask(event.state()) != null) {
+                clusterService.removeListener(taskStarter);
+            } else if (event.localNodeMaster()) {
+                clusterService.removeListener(taskStarter);
```

Comment:
it seems we're removing the task starter on both branches of this if/else clause. Could this be simplified?

Reply:
I gave it a shot, let me know if it's clear as well as simplified.